### PR TITLE
Main candidate from NCAR (04/21/2021)

### PR DESCRIPTION
**Pull request by gustavo-marques**
### Summary 

* Many improvements in the lateral boundary diffusion module 

* Adds a new counter-based random number generator

* Fixes a minor bug in the neutral diffusion module (only applies when **NDIFF_INTERIOR_ONLY = True**)

* Fixes sign bug in the latent heat contribution from frozen precipitation and frozen runoff (MCT and NUOPC caps)

* Adds option to apply an are flux correction in the NUOPC cap

* Adds support for threading in the NUOPC cap

This PR does not change the MOM6-examples answers (w.r.t. branch main at 1a420f2) when using the intel/19.0.5 compiler. It also does not add new MOM input flags or diagnostics.

### CODE REVIEW

**Comment by gustavo-marques** [comment]
@kshedstrom, @jiandewang, @abozec (or @awallcraft): please review/test this PR.

**Comment by jiandewang** [comment]
@gustavo-marques https://github.com/NOAA-GFDL/MOM6/pull/1381/files#diff-68076a5015442fbab97fced84c6d2c139c715798d26b7e0fe0c542bab0b3c1d6R467-R478
use function ChkErr(rc,__LINE__,u_FILE_u) to replace 
if (ESMF_LogFoundError

**Comment by gustavo-marques** [comment]
@jiandewang, thanks for catching that.

**Comment by kshedstrom** [comment]
I approve this PR.

**Review by abozec — APPROVED** [review]
COAPS approves this PR.

File: config_src/drivers/nuopc_cap/mom_cap.F90
Comment (DeniseWorthen) [inline config_src/drivers/nuopc_cap/mom_cap.F90:469]: The two instances of retrieving the attribute 'nthreads' are failing for EMC. They will need "isPresent,isSet" conditionals added. Thanks.

Comment (gustavo-marques) [reply config_src/drivers/nuopc_cap/mom_cap.F90:469]: Done. Thanks, @DeniseWorthen.

**Comment by DeniseWorthen** [comment]
@gustavo-marques This still isn't working for EMC. We need to use the isPresent,isSet conditionals to optionally retrieve the config variable. I've tested the following [change](https://github.com/gustavo-marques/MOM6/compare/main-candidate-ncar-2021-04-21...DeniseWorthen:feature/nthrd_config) for our configuration on Gaea: 

Please check the logic and make sure it is what you intend. Thanks.

```
diff --git a/config_src/drivers/nuopc_cap/mom_cap.F90 b/config_src/drivers/nuopc_cap/mom_cap.F90
index f5f7985e1..742b7575d 100644
--- a/config_src/drivers/nuopc_cap/mom_cap.F90
+++ b/config_src/drivers/nuopc_cap/mom_cap.F90
@@ -418,6 +418,7 @@ subroutine InitializeAdvertise(gcomp, importState, exportState, clock, rc)
   character(len=512)                     :: diro
   character(len=512)                     :: logfile
   character(ESMF_MAXSTR)                 :: cvalue
+  character(len=64)                      :: logmsg
   logical                                :: isPresent, isPresentDiro, isPresentLogfile, isSet
   logical                                :: existflag
   integer                                :: userRc
@@ -467,13 +468,19 @@ subroutine InitializeAdvertise(gcomp, importState, exportState, clock, rc)
   if (ChkErr(rc,__LINE__,u_FILE_u)) return

   if(localPeCount == 1) then
-     call NUOPC_CompAttributeGet(gcomp, "nthreads", value=cvalue, &
+     call NUOPC_CompAttributeGet(gcomp, name="nthreads", value=cvalue, &
           isPresent=isPresent, isSet=isSet, rc=rc)
      if (ChkErr(rc,__LINE__,u_FILE_u)) return
-     read(cvalue,*) nthrds
+     if (isPresent .and. isSet) then
+        read(cvalue,*) nthrds
+     else
+        nthrds = localPeCount
+     end if
   else
      nthrds = localPeCount
   endif
+  write(logmsg,*) nthrds
+  call ESMF_LogWrite(trim(subname)//': nthreads = '//trim(logmsg), ESMF_LOGMSG_INFO)

 !$  call omp_set_num_threads(nthrds)

@@ -898,10 +905,14 @@ subroutine InitializeRealize(gcomp, importState, exportState, clock, rc)
   if (ChkErr(rc,__LINE__,u_FILE_u)) return

   if(localPeCount == 1) then
-     call NUOPC_CompAttributeGet(gcomp, "nthreads", value=cvalue, &
+     call NUOPC_CompAttributeGet(gcomp, name="nthreads", value=cvalue, &
           isPresent=isPresent, isSet=isSet, rc=rc)
      if (ChkErr(rc,__LINE__,u_FILE_u)) return
-     read(cvalue,*) nthrds
+     if (isPresent .and. isSet) then
+        read(cvalue,*) nthrds
+     else
+        nthrds = localPeCount
+     end if
   else
      nthrds = localPeCount
   endif
```

**Comment by gustavo-marques** [comment]
@DeniseWorthen, sorry for breaking your tests, and thank you for suggesting a solution. I will add your changes and run them by Jim Edwards since he implemented the threading option in NUOPC.

**Comment by jiandewang** [comment]
@gustavo-marques I will run UFS RT with your updated code, but may take some time as we ran out of CPU quota at this moment, job turnover rate is extremely low.

**Comment by jiandewang** [comment]
@gustavo-marques: by adding !$  call omp_set_num_threads(nthrds) in the cap code, does it mean part of cap will be running with OMP or something else ? we don't have a clear idea here, can you explain ?

The updated code passed UFS RT, thanks for @DeniseWorthen and @gustavo-marques

**Comment by alperaltuntas** [comment]
@jiandewang, these (or any) OpenMP directives are effective only if you explicitly enable OpenMP at compile time. Unless you compile with OpenMP, these directives (!$) are merely regarded as comments, and so will not affect your model. If you are compiling with OpenMP, on the other hand, the directive you refer to does the following: "Instead of using the default number of threads, the directive sets the number of threads to the value received from NUOPC_CompAttributeGet call.

Also mentioned @jedwards4b, if he likes to add more.

**Comment by jiandewang** [comment]
> @jiandewang, these (or any) OpenMP directives are effective only if you explicitly enable OpenMP at compile time. Unless you compile with OpenMP, these directives (!$) are merely regarded as comments, and so will not affect your model. If you are compiling with OpenMP, on the other hand, the directive you refer to does the following: "Instead of using the default number of threads, the directive sets the number of threads to the value received from NUOPC_CompAttributeGet call.
> 
> Also mentioned @jedwards4b, if he likes to add more.

@alperaltuntas thanks for the explanation, did you get identical results from different threading # ?

**Comment by marshallward** [comment]
GFDL Gaea regression (internal): https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/12566

**Comment by alperaltuntas** [comment]
@jiandewang Yes, the number of threads doesn't change the results.

**Comment by marshallward** [comment]
@sanAkel Did you want to review this PR?  The number of non-NUOPC changes is very small so I would not expect any issues.

Otherwise, I will go ahead and merge this if there's no objection.  (I will redirect it to `main` rather than `dev/gfdl`).

**Comment by sanAkel** [comment]
> @sanAkel Did you want to review this PR? The number of non-NUOPC changes is very small so I would not expect any issues.
> 
> Otherwise, I will go ahead and merge this if there's no objection. (I will redirect it to `main` rather than `dev/gfdl`).

@marshallward Thanks for checking. I think I will take a pass on this one. Though I am a bit confused by "(I will redirect it to `main` rather than `dev/gfdl`)" You won't merge it to `dev/gfdl` ?

**Comment by marshallward** [comment]
> I am a bit confused by "(I will redirect it to `main` rather than `dev/gfdl`)" You won't merge it to `dev/gfdl` ?

The PR was misdirected to the `dev/gfdl` branch, but we want it to go to `main`, i.e. our release branch.  I will manually merge this to `main` but will later merge the contents into `dev/gfdl`.

**Comment by sanAkel** [comment]
> > I am a bit confused by "(I will redirect it to `main` rather than `dev/gfdl`)" You won't merge it to `dev/gfdl` ?
> 
> The PR was misdirected to the `dev/gfdl` branch, but we want it to go to `main`, i.e. our release branch. I will manually merge this to `main` but will later merge the contents into `dev/gfdl`.

Got it, Thanks @marshallward. Once you have
> later merge the contents into `dev/gfdl`.

We will update GEOS-ESM.

**Comment by marshallward** [comment]
This PR has now been added to `main`.  It should auto-close once we merge into `dev/gfdl`.

**Pull request merged**
